### PR TITLE
Add cephfs extras endpoints

### LINKED ISSUE #405: Add API endpoints for handling /extras fileshare contents
### Description of the issue

The FIA API needs endpoints to interact with the cephFS mount. This will enable the frontend to implement its upload feature (previously intended for the admin-portal)

### PR DESCRIPTION
**Pull request by HumzahJavid**
Closes #405 

The FIA API needs additional endpoints to handle interactions with the cephFS mount
This PR will add the following endpoints and respective tests:
- `GET /extras` to view all the instrument subfolders instrument folder
- `GET /extras/<instrument>`
- `PUT /extras/<instrument>/<filename>`

This PR will add:
- utility functions and respective tests

### CODE REVIEW

File: fia_api/core/file_ops.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+
+def read_dir(path: Path) -> list[str]:
+    """
+    Read the names of each file in the given path
+    :param path: The path to the directory to get the filenames of
+    :return: A list of paths in the given param path
+    """
+    try:
+        return os.listdir(path)
```
[the remark this answers is not in the document]
Comment (HumzahJavid) [reply fia_api/core/file_ops.py:18]: Overall a false positive, as the `safe_check_filepath` function is called before this. Added another instance for checking a path is safe against itself (for completions sake, to sanitise the input)

File: fia_api/core/file_ops.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        ) from err
+
+
+async def write_file_from_remote(remote_file: UploadFile, local_file: Path) -> None:
+    """
+    Write the contents of the remote_file from called to the local path given.
+    """
+    try:
+        contents = await remote_file.read()
+        path = anyio.Path(local_file)
```
[the remark this answers is not in the document]
Comment (HumzahJavid) [reply fia_api/core/file_ops.py:32]: A false positive as the safe_check_filepath function is called before this.

File: fia_api/core/utility.py [hunk 3]
```diff
@@ -34,3 +38,22 @@
...
+
+def safe_check_filepath(filepath: Path, base_path: Path) -> None:
+    """
+    Check to ensure the path does contain the base path and that it does not resolve to some other directory
+    :param filepath: the filepath to check
+    :param base_path: base path to check against
+    :return:
+    """
+    try:
+        filepath.resolve(strict=True)
```
[the remark this answers is not in the document]
Comment (HumzahJavid) [reply fia_api/core/utility.py:51]: False positive, this function solves the problem being flagged.

File: .vscode/launch.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (Pasarus) [inline .vscode/launch.json:1]: Please add .vscode to the .gitignore file.

File: fia_api/core/utility.py [hunk 5]
```diff
@@ -34,3 +38,22 @@ def filter_script_for_tokens(script: str) -> str:
...
+        if not filepath.is_relative_to(base_path):
+            raise HTTPException(status_code=HTTPStatus.FORBIDDEN, detail="Invalid path being accessed.")
+    except FileNotFoundError as err:
```
Comment (Pasarus) [inline fia_api/core/utility.py:54]: Would it be possible to test this error does infact get raised? and that the subsequent execution branches are covered in tests.

File: fia_api/router.py [hunk 6]
```diff
@@ -290,3 +294,52 @@ async def update_instrument_specification(
     """
     update_specification_for_instrument(instrument_name.upper(), specification)
     return specification
+
+
+@ROUTER.get("/extras", tags=["files"])
+async def get_extras_top_level_folders() -> list[str]:
```
Comment (Pasarus) [inline fia_api/router.py:300]: Are there e2e tests for /extras the endpoint?

File: fia_api/router.py [hunk 7]
```diff
@@ -290,3 +294,52 @@ async def update_instrument_specification(
...
+    \f
+    :return: List of folders
+    """
+    root_directory = Path(os.environ.get("EXTRAS_DIRECTORY", "/extras"))
+    safe_check_filepath(root_directory, root_directory)
+    return read_dir(root_directory)
+
+
+@ROUTER.get("/extras/{subdir}", tags=["files"])
+async def get_subfolder_files_list(subdir: str) -> list[Path]:
```
Comment (Pasarus) [inline fia_api/router.py:312]: Are there e2e tests for /extras/{} the endpoint?

File: fia_api/router.py [hunk 8]
```diff
@@ -290,3 +294,52 @@ async def update_instrument_specification(
...
+    """
+    root_directory = Path(os.environ.get("EXTRAS_DIRECTORY", "/extras"))
+    subdir_path = root_directory / subdir
+    safe_check_filepath(subdir_path, root_directory)
+
+    return read_dir(subdir_path)
+
+
+@ROUTER.post("/extras/{instrument}/{filename}", tags=["files"])
+async def upload_file_to_instrument_folder(instrument: str, filename: str, file: UploadFile) -> str:
```
Comment (Pasarus) [inline fia_api/router.py:327]: Are there e2e tests for /extras/{}/{} the endpoint?

**Review by Pasarus — CHANGES_REQUESTED** [review]
Just a couple initial questions

**Comment by Pasarus** [comment]
You also seem to be failing on formatting and linting: https://github.com/fiaisis/FIA-API/actions/runs/11912167155/job/33195191453?pr=406

File: fia_api/router.py (hunk 8 above)
Comment (HumzahJavid) [reply fia_api/router.py:327]: Yes, https://github.com/fiaisis/FIA-API/blob/7c1d39051b49d0482fd2411c9ecc113a3ec5f7c6/test/e2e/test_file_upload.py#L117-L144 

But if all these tests were not easily identifiable, then maybe the test file and tests need a rename/restructure?

**Review by Pasarus — COMMENTED** [review]
There seems to be some issues with coverage on the e2e tests not being tracked so what I will do is make an issue for it in the backlog

File: fia_api/core/utility.py (hunk 5 above)
Comment (Pasarus) [reply fia_api/core/utility.py:54]: Can we have a explicit test?

**Review by Pasarus — APPROVED** [review]

**Pull request merged**